### DISCUSSION TITLE
Counters & variables - my or Tasmotas problem?

### DISCUSSION THREAD

**Discussion by Pewibe** [arendst/Tasmota#17089] [Q&A] [answered]
In conjunction with gas meter reading/counting I discovered some problems where I don't know if it is my wrong usage (lack of documentation reading) or Tasmotas wrong/problematic function.

1.  Wrong calculation 17345678 + 1 = 17345680
16:54:25.095 CMD: var5 17345678
16:54:25.102 MQT: stat/H.GasZaehler/RESULT = {"Var5":"17345678"}
16:54:29.057 CMD: add5 1
16:54:29.064 MQT: stat/H.GasZaehler/RESULT = {"Add5":"**17345680**.000"}
Documentation says "The value of a Var<x> can be any number". This means the value of the above example is not too high. Additionally doc says "single precision" which means the value is too high. What to do if the calculation value is derived from a counter 32 bit like my example?

2. Counter setting/increment with variable
Documentation says "1.. = preset, -1.. = decrease, +1.. = increase Counter". If I want to use a variable instead of a constant (Counter1 %Var2%), how is it distinct between preset and increase?

3. Counter reading in rule
I like to assign a counter to a variable and tried "Var1 %Counter2%" which didn't work. Any hint?

4. Counter name
Is a counters name "Zaehler1" changeable to "Something" without editing/compilation?

**Comment by sfromis** [comment, marked as answer]
1) Arithmetic in rules is indeed single precision floating point, basically about 7 digits, over that you lose right-order precision. "Any number" does not mean full precision. In rules you cannot have calculations with more digits.

2) In rules you can use a variable anywhere in a command using the notation %var1%
Using counters is also a usable workaround to add/subtract 31-bit integers, up to about 2 billion.

3) You cannot "read" counters in a rule, what you can do is to use a trigger to react to the counter being updated, and then do something with the value.

4) No feature to change names of counters

**Reply by Pewibe** [reply]
Answer 1, 3, 4 are OK. Concerning 2 I think you misunderstood:
In a rule Counter1 should be incremented by 1 or 2. The related increment is contained in Var2 but rule "do counter1 %var2% endon" want work. Any idea to increment a counter by variable Value?

**Reply by sfromis** [reply]
I understood what you wrote, and answered to that. If you have var2=2 and want the command `Counter1 +2`, then your rule can contain `Counter1 +%var2%`

And this way you can do integer add/subtract with full precision integers up to 2 billion (9 digits and a bit more).

**Reply by Pewibe** [reply]
Thanks, this construction "+%var2%" was the hint I needed!
